### PR TITLE
Refactors embedding, adds harmless embeds, sticky tape, admin buttons to make everything embeddable

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
Take 2! Was unable to reopen the last one since I forcepushed while it was closed, but you can find the last PR at #47861

[](https://i.imgur.com/VWBcEe1.png)
_Everything is pointy here because of the "make everything embeddable" button_

[

_Unused, sticky tape, super sticky tape, pointy tape, super pointy tape_
Credits to MrDoomBringer for tape sprites from hippie, and Vekter for the idea.

Features:
- Adds harmless embeds: Items with no pain chance are considered harmless, and skip wounding (no damage, no blood, no slice sound, etc), as well as being described as sticking to people rather than being embedded in them. Toy shurikens are added to show this off.
- Embeds can split between brute and stamina damage: Items can deal a percentage of their damage as stamina rather than (or instead of) just brute. Included are shock shurikens, described more in the next point.
- Embeds can jostle on movement: As well as randomly causing pain while stuck in someone, embeds can cause pain randomly as the target moves around. Walking or crawling procs jostling less often than running, meaning jostling is more effective on a fleeing target than a stationary one. Shock shurikens deal mostly stamina damage, and mostly through jostling, making them a delayed alternative to bolas for less lethal takedowns.
- Adds sticky tape, an item that can be used to make anything a harmless embed. Stick a can of monkey energy to the captain's head from across the hall! Plant a banana peel on the mime's back to show him who's boss! All new possibilities for harassment. There's also a version that lets anything become a harmful embeddable, but I think that'll stay admin only.
- Adds buttons to make everything sticky/embeddable: A button to make everything a harmless embeddable, and another button to make everything a harmful embeddable. I did this as an event once during a wizard round and watching assistants running around sticking action figures and medkits in each other was a boatload of fun. With maintainer permission, I'd like to add these to wizard summon events since summon events is in dire need of more variety.

I'd like thoughts on shock shurikens and sticky tape. I envision shock shurikens being (perhaps premium) sec equipment similar to bolas, but I don't know if they'd make more sense in the sec vendor, a cargo crate, or a techfab design. Honestly I made it as a showcase for stamina/jostling more than I did as a balanced, competitive sec tool. As for sticky tape, I think the B.E.P.I.S. would make a good home for it, since it's dumb harmless fun. Toy shurikens will probably get added to maint loot or arcade machines.

Also still have a little bit of code neatening and standardizing to finish up.

## Why It's Good For The Game
New behavior for embeddable objects that lets them be used in new ways (less lethal takedowns, pranks and hijinx) gives players and developers more choices for how to use them, and sticky tape and buttons to make everything sticky/embeddable is funny. Also, centralizes embedding code to mostly one file.

## Changelog
:cl: Ryll/Shaps
tweak: Embeddables have been reworked and given new functionality. 
add: Adds toy shurikens for pranks and shock shurikens for less lethal takedowns.
add: Adds sticky tape that lets you stick random junk to people! How will anyone believe the Captain's cries that he's not actually a boomer when there's a can of Monkey Energy stuck to his head?
refactor: Embedding functionality has been made an element, with new support for dealing a mix of brute and stamina damage, and a separate pain check for on-move.
/:cl:

### CODE REVIEW

**Comment by SkeletalElite** [comment]
I really like the shock shurikens. It's a cool alternative to the energy bola, how good it is will depend on how long it takes to pull it out. If it takes a little longer than the energy bola I think that's a good way for it. 
That way the energy bola is more effective at quickly taking someone down while theyre slow but is quick to remove, but the shuriken is good at widdling a target down. Keeping them from getting them out but with less immediate benefit compared to the bola.

**Comment by Michael-Ashfield** [comment]
New meta, tape the nuke disk to the captain so it is never dropped

**Comment by evilcyborgmrt** [comment]
Can't wait to hide in plants and stick rotting gibs at unsuspecting passers by. Or severed cat tails! Oh god so many pranks can ensue from this plz merge

**Comment by Tlaltecuhtli** [comment]
can you make embedding check for armor stats (better melee armor > less chance of embed)?

**Comment by necromanceranne** [comment]
Sticky tape?

Sticky grenade.

**Comment by Ryll-Ryll** [comment]
Good point, I'll give grenades a 50% chance to get stuck to your hand when thrown and make you have to pry it off

**Comment by blessedmulligan** [comment]
good pr because it taught me that throwing stars do a ton of damage

**Comment by blessedmulligan** [comment]
Gonna nitpick on an incidental feature added by this PR, but a points on shock stars: First of all, they probably need a different sprite. 

After an extensive 5 minutes of testing, I've come to the conclusion that they're pretty unreliable. To actually stop your target, you'll probably need to land two stars, since there's a good chance the first star hits a limb, which caps the damage it can do. Crippling your target's leg or arm is still good, but that takes a while. With two stars stuck, you have a reasonably quick knockdown if they keep running, but it still feels pretty RNG reliant as to exactly how long it takes.

**Comment by Shadowflame909** [comment]
Don't forget e-swords are the most powerful embedding weapon.

People just don't throw them because you lose your E-sword.

**Comment by blessedmulligan** [comment]
Not sure if this is added by this PR because I'm too lazy to test it in the main branch but ahealing someone with embedded objects removes the objects themselves but doesn't stop the DoT

**Comment by necromanceranne** [comment]
@Shadowflame909 actually I'm pretty sure eswords and edaggers do the same damage on embed. Worth it for edaggers they're 2tc a pop.

**Comment by ATH1909** [comment]
I know how to pull an embedded item out of/off of yourself, but how do you remove an embedded item from another person? This could become relevant if assistants start embedding tooloxes in themselves in order to get a storage item that's hard for sec to remove from you.

**Comment by TheVekter** [comment]
Credit me for the idea you fack

**Comment by HomocidlePancakes** [comment]
I used to steal shoes from sec.
Now I will stick the shoes to their foreheads after stealing them.

**Comment by spookydonut** [comment]
https://github.com/tgstation/TerraGov-Marine-Corps/pull/2924

**Comment by Ryll-Ryll** [comment]
I'm not sure what you mean by linking that PR considering there's 56 changed files and the code paradigms are totally different

Also yes the original idea for harmless embeds was courtesy of vekter, I'll update the OP!

You can remove embeds from other people with a 2 step surgery, just needs drapes, a scalpel, and a hemostat I believe

Also oranges denied the wizard events pass so I'll disable them tonight

**Comment by Anonmare** [comment]
Old fashioned electrode tasers when?

Also were gonna need Bay's yank out object verb at some point

**Comment by blessedmulligan** [comment]
If you dismember someone, you can cut open their limb with a sharp object and remove any embedded objects. This is clearly the only appropriate response for security to have when they arrest an assistant with soap stuck to his leg.

**Comment by Ryll-Ryll** [comment]
Why are you saying sticky objects don't have a random fall chance when they do? You don't need surgery to take sticky soap off someone's leg, it will fall off. As for stun stars, they have greatly reduced damage from normal random pain. If you stand still you would definitely outheal the stamina pain.

**Comment by itseasytosee** [comment]
It's back. I've been waiting

**Comment by blessedmulligan** [comment]
>There is literally 0 risk of permastunning people before they can rip it out

This is definitely much less of a problem than I thought, so I apologize for that. Still, my understanding, which is supported by some quick testing, is that stamina damage lasts until no stamina damage is taken for several seconds. So if someone goes into stamcrit from the shock stars (which they would have to be pretty dumb to let happen but it could), they'll stay in stamcrit until they go long enough between ticks for the stamina damage to go away. Since the ticks are rng, this could be indefinitely, although it does usually clear pretty quickly (depending on how many stars they have in them...)

I also forgot about the fact that they still do actual damage, so in the worst case it will eventually kill them (from either bleeding or brute damage) instead of keeping them at 100% health but indefinitely incapacitated.

I still think having a way to pull things out by hand instead of just waiting or doing surgery is better.

**Comment by SkeletalElite** [comment]
I'm not certain how often pain ticks are but if they're 2 seconds like life then there's roughly 20% chance someone in stam crit will get their stun extended by the stamina damage. And each time it extends there's an additional 20% chance that you get your stam crit reset again. So it is possible to get really unlucky and end up with a really long stun but you have to be unlucky. 
That percentage is a tiny bit higher if the 5th tick occurs before the stamina damage purge, not sure which will happen first. It's a tiny bit lower if that isn't the case, +-2%

File: code/datums/elements/embed.dm [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		return ELEMENT_INCOMPATIBLE
+
+	RegisterSignal(target, COMSIG_MOVABLE_IMPACT_ZONE, .proc/embed_check)
+	RegisterSignal(target, COMSIG_ITEM_EMBED_REMOVING_RIP, .proc/rip_out)
+	RegisterSignal(target, COMSIG_ITEM_EMBED_REMOVE_SURGERY, .proc/safe_remove)
+
+/datum/element/embed/Detach(obj/item/target)
+	. = ..()
+
+	if(target in active_embeds)
```
Comment (ninjanomnom) [inline code/datums/elements/embed.dm]: You don't need the if, just remove it

File: code/datums/elements/embed.dm [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		active_embeds -= target
+
+	UnregisterSignal(target, COMSIG_MOVABLE_IMPACT_ZONE)
+	UnregisterSignal(target, COMSIG_ITEM_EMBED_REMOVING_RIP)
+	UnregisterSignal(target, COMSIG_ITEM_EMBED_REMOVE_SURGERY)
+
+/datum/element/embed/proc/jostle_check(mob/living/carbon/human/victim)
+	for(var/obj/item/bodypart/L in victim.bodyparts)
+		for(var/obj/item/weapon in L.embedded_objects)
+			var/datum/embedding_behavior/EB = weapon.embedding
```
Comment (ninjanomnom) [inline code/datums/elements/embed.dm]: The embedding behavior should be in this element

File: code/datums/elements/embed.dm [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+			if(!weapon.is_embed_harmless())
+				var/chance = EB.jostle_chance
+				if(victim.m_intent == MOVE_INTENT_WALK || victim.lying)
+					chance *= 0.5
+
+				if(prob(chance))
+					var/damage = weapon.w_class * EB.jostle_pain_mult
+					L.receive_damage(brute=(1-EB.pain_stam_pct) * damage, stamina=EB.pain_stam_pct * damage)
+					to_chat(victim, "<span class='userdanger'>[weapon] embedded in your [L.name] jostles and stings!</span>")
```
Comment (ninjanomnom) [inline code/datums/elements/embed.dm]: Each weapon's embed element should be handling this behavior

File: code/datums/elements/embed.dm [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	weapon.forceMove(get_turf(victim))
+	active_embeds -= weapon
+
+	if(!victim.has_embedded_objects())
+		victim.clear_alert("embeddedobject")
+		UnregisterSignal(victim, COMSIG_MOVABLE_MOVED)
+		SEND_SIGNAL(victim, COMSIG_CLEAR_MOOD_EVENT, "embedded")
+
+/datum/element/embed/process()
+	for(var/obj/item/weapon in active_embeds)
```
Comment (ninjanomnom) [inline code/datums/elements/embed.dm]: If you know that only weapons will be in the list you don't need a filtered loop

File: code/datums/elements/embed.dm [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		var/datum/embedding_behavior/EB = weapon.embedding
+		var/obj/item/bodypart/L = weapon.embedded_in
+		var/mob/living/carbon/human/victim = L.owner
+
+		if(!victim || !L) // in case the victim and/or their limbs exploded (say, due to a sticky bomb)
+			active_embeds -= weapon
+			if(victim && !victim.has_embedded_objects())
+				UnregisterSignal(victim, COMSIG_MOVABLE_MOVED)
+			weapon.embedded_in = null
```
Comment (ninjanomnom) [inline code/datums/elements/embed.dm]: What with you needing to process and keeping track of embed state in all sorts of place, you should probably not be trying to do this all from a single element. Discuss this with me on discord

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by Ryll-Ryll** [comment]
:)

File: code/datums/components/embeds/embedded_structure.dm [hunk 6]
```diff
@@ -0,0 +1,67 @@
+/datum/component/embedded_structure
```
Comment (ninjanomnom) [inline code/datums/components/embeds/embedded_structure.dm]: You should only need a single component for the various types

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/embeds/embedded_structure.dm (hunk 6 above)
Comment (Ryll-Ryll) [reply code/datums/components/embeds/embedded_structure.dm]: Woops, I just have them separate right now for testing purposes, hence the last commit name

File: code/game/objects/items.dm [hunk 7]
```diff
@@ -149,12 +152,23 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
...
-	if (!embedding)
-		embedding = getEmbeddingBehavior()
-	else if (islist(embedding))
-		embedding = getEmbeddingBehavior(arglist(embedding))
-	else if (!istype(embedding, /datum/embedding_behavior))
-		stack_trace("Invalid type [embedding.type] found in .embedding during /obj/item Initialize()")
+	if(embedding)
+		var/list/temp = embedding
+		temp.Insert(0, /datum/element/embed) // i dunno about this
+		AddElement(arglist(temp))
```
Comment (AnturK) [inline code/game/objects/items.dm]: You could just pass the parameters as list, these are unique by id so bespoke will get same ref anyway.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/datums/components/embedded.dm [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+
+/datum/component/embedded/RegisterWithParent()
+	if(ishuman(parent))
+		RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/jostleCheck)
+		RegisterSignal(parent, COMSIG_HUMAN_EMBED_RIP, .proc/ripOutHuman)
+		RegisterSignal(parent, COMSIG_HUMAN_EMBED_REMOVAL, .proc/safeRemoveHuman)
+	else if(isturf(parent))
+		RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examineTurf)
+		RegisterSignal(parent, COMSIG_TOPIC, .proc/ripOutTurf)
```
[the remark this answers is not in the document]
Comment (Ryll-Ryll) [reply code/datums/components/embedded.dm]: This is just a continuation of how embedding has always worked, where you inspect yourself and you get links in your chat that you can click on to get them unstuck, and I'm under the impression chat links have to go through topic as that's how they work. It's a bit dated, yeah, but I haven't been able to think of a better way of viewing and interacting with stuff stuck inside you that would scale to having upwards of a dozen instances at once.

Unless you're saying there's a way to do chat links without topic, in which case, I'm interested!

[the remark this answers is not in the document]
Comment (AnturK) [reply code/datums/components/embedded.dm]: What i'm saying is this should simply be component Topic not this signal.

File: code/datums/components/embedded.dm [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+	src.rip_time = rip_time
+	src.ignore_throwspeed_threshold = ignore_throwspeed_threshold
+	src.jostle_chance = jostle_chance
+	src.jostle_pain_mult = jostle_pain_mult
+	src.pain_stam_pct = pain_stam_pct
+
+	src.weapon = I
+
+	if(src.pain_chance || src.jostle_chance)
+		harmful = TRUE
```
Comment (Qustinnus) [inline code/datums/components/embedded.dm]: Does this need a var? Sounds like a simple helper works

File: code/datums/components/embedded.dm [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+
+/datum/component/embedded/RegisterWithParent()
+	if(ishuman(parent))
+		RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/jostleCheck)
+		RegisterSignal(parent, COMSIG_HUMAN_EMBED_RIP, .proc/ripOutHuman)
+		RegisterSignal(parent, COMSIG_HUMAN_EMBED_REMOVAL, .proc/safeRemoveHuman)
+	else if(isturf(parent))
+		RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examineTurf)
+		RegisterSignal(parent, COMSIG_TOPIC, .proc/ripOutTurf)
+		//RegisterSignal(parent, COMSIG_PARENT_QDELETING, .proc/destroyedTurf)
```
Comment (Qustinnus) [inline code/datums/components/embedded.dm]: Remove unused code

File: code/datums/components/embedded.dm [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+	weapon.forceMove(victim)
+
+	if(harmful)
+		victim.visible_message("<span class='danger'>[weapon] embeds itself in [victim]'s [L.name]!</span>","<span class='userdanger'>[weapon] embeds itself in your [L.name]!</span>")
+		victim.throw_alert("embeddedobject", /obj/screen/alert/embeddedobject)
+		playsound(victim,'sound/weapons/bladeslice.ogg', 40)
+		weapon.add_mob_blood(victim)//it embedded itself in you, of course it's bloody!
+		var/damage = weapon.w_class * impact_pain_mult
+		L.receive_damage(brute=(1-pain_stam_pct) * damage, stamina=pain_stam_pct * damage)
+		SEND_SIGNAL(victim, COMSIG_ADD_MOOD_EVENT, "embedded", /datum/mood_event/embedded)
```
Comment (Qustinnus) [inline code/datums/components/embedded.dm]: Maybe a small moodlet for getting something harmless stu k on you might be fun

File: code/datums/elements/embed.dm [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+				remove_pain_mult = remove_pain_mult,\
+				rip_time = rip_time,\
+				ignore_throwspeed_threshold = ignore_throwspeed_threshold,\
+				jostle_chance = jostle_chance,\
+				jostle_pain_mult = jostle_pain_mult,\
+				pain_stam_pct = pain_stam_pct)
+
+
+/// We need the hit_zone if we're embedding into a human, so this proc only handled if we're embedding into a turf
+/datum/element/embed/proc/checkEmbedOther(obj/item/weapon, turf/hit, datum/thrownthing/throwingdatum)
```
Comment (Qustinnus) [inline code/datums/elements/embed.dm]: Like I said in discord, being able to embed in turfs should be both a Turf and item flag.. Throwing floor tiles and shards into walls sounds odd

File: code/game/objects/items.dm [hunk 13]
```diff
@@ -222,6 +229,12 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 		if(resistance_flags & FIRE_PROOF)
 			. += "[src] is made of fire-retardant materials."
 
+	if(embedding)
```
Comment (Qustinnus) [inline code/game/objects/items.dm]: I'm phone posting but is this an examine? Yiu should put this in the element

**Review by ghost — CHANGES_REQUESTED** [review]
Got tired will review more tommorow good progress so far

**Comment by Ryll-Ryll** [comment]
Work on tackling went long tonight, I'll fix up the last pieces of this during this weekend

**Comment by cacogen** [comment]
fun!

File: code/datums/elements/embed.dm (hunk 12 above)
Comment (Ryll-Ryll) [reply code/datums/elements/embed.dm]: Turns out walls already have a hardness var for hulk smashing and such, I made normal walls have a +0 modifier, and any harder/softer results in a modifier to the embed_chance roll

File: code/datums/components/embedded.dm (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ryll-Ryll) [reply code/datums/components/embedded.dm]: done in latest commit

File: code/datums/components/embedded.dm (hunk 9 above)
Comment (Ryll-Ryll) [reply code/datums/components/embedded.dm]: well it's not like it'll ever change during the component's life, so one-and-done is probably better as a micro-opt

File: code/game/objects/items.dm (hunk 13 above)
Comment (Ryll-Ryll) [reply code/game/objects/items.dm]: bleh now it's all gonna be out of order but oh well!

**Comment by Ryll-Ryll** [comment]
I added sticky tape to maint loot, though you'll still need the BEPIS reward to print them out on demand

**Comment by spookydonut** [comment]
cursed title

**Comment by Ryll-Ryll** [comment]
The sooner this gets merged the sooner you'll never have to see it again

**Comment by cacogen** [comment]
so it begins

**Pull request merged**